src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+            return;
+        }
+        checkIndex(index, "Cannot add value by index: ");
+        Node<T> node;
+        if (index == ZERO) {
+            node = new Node<>(value, null, first);
+            first.prev = node;
+            first = node;
+        } else {
+            Node<T> nodeAtIndex = findNodeByIndex(index);
+            node = new Node<>(value, nodeAtIndex.prev, nodeAtIndex);
+            nodeAtIndex.prev.next = node;
+            nodeAtIndex.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, "Cannot get the value by index: ");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, "Cannot set value by index: ");
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        T replacedValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, "Cannot remove value by index: ");
+        Node<T> node = first;
+        if (index == ZERO) {
+            removeFirst();
+            return node.value;
+        }
+        if (index == size - ONE) {
+            node = last;
+            removeLast();
+            return node.value;
+        }
+        node = findNodeByIndex(index);
+        unlink(node);
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> nodeToRemove = findNodeByObject(object);
+        if (nodeToRemove == null) {
+            return false;
+        }
+        if (nodeToRemove.prev == null) {
+            removeFirst();
+            return true;
+        }
+        if (nodeToRemove.next == null) {
+            removeLast();
+            return true;
+        }
+        unlink(nodeToRemove);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == ZERO;
+    }
+
+    private void checkIndex(int index, String message) {
+        if (index >= size || index < ZERO) {
+            throw new IndexOutOfBoundsException(message
+                    + "Index is out of bounds");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        int listHalf = size / TWO;
+        if (index < listHalf) {
+            Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        }
+        Node<T> node = last;
+        for (int i = size - ONE; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
```

Comment:
try to simplify and use only one return statement (just use if-else and one return after "else" block)